src/compiler/dataStructures.ts

```diff
@@ -0,0 +1,777 @@
+            return (<any>map).size;
+        }
+        else {
+            let size = 0;
+            map.forEach(() => { size++; });
+            return size;
+        }
+    }
+
+    /** Iterate through a map, returning the first truthy value returned by `fn`, or undefined. */
+    export function findInMap<V, U>(map: SMap<V>, fn: (value: V, key: string) => U | undefined): U | undefined {
+        if (map instanceof Map) {
+            // Using an iterator and testing for `done` performs better than using forEach() and throwing an exception.
+            const iterator: { next(): { value: [string, V], done: boolean } } = (<any>map).entries();
+            while (true) {
+                const { value: pair, done } = iterator.next();
+                if (done) {
+                    return undefined;
+                }
+                const [key, value] = pair;
+                const result = fn(value, key);
+                if (result) {
+                    return result;
+                }
+            }
+        }
+        else {
+            return (<ShimSMap<V>>map).find(fn);
+        }
+    }
+
+    /** Iterate over every key. */
+    export function forEachKeyInMap<V>(map: SMap<V>, callback: (key: string) => void): void {
+        map.forEach((_, key) => callback(key));
+    }
+
+    /** Copy all entries from `source` to `target`, overwriting any already existing. */
+    export function copyMap<V>(source: SMap<V>, target: SMap<V>): void {
+        source.forEach((value, key) => {
+            target.set(key, value);
+        });
+    }
+
+    /** Array of all keys in a map. */
+    export function keysArray<V>(map: SMap<V>): string[] {
+        const keys: string[] = [];
+        forEachKeyInMap(map, key => keys.push(key));
+        return keys;
+    }
+
+    /** Array of all values in a map. */
+    export function valuesArray<K, V>(map: SMap<V>): V[] {
+        const values: V[] = [];
+        map.forEach(value => values.push(value));
+        return values;
+    }
+
+    /** Set `map.get(key) === value`, and return `value`. */
+    export function setAndReturn<K, V>(map: MapCommon<K, V>, key: K, value: V): V {
+        map.set(key, value);
+        return value;
+    }
+
+    /** Like `getOrUpdateMap` but does not return a value. */
+    export function setButDontOverride<K, V>(map: MapCommon<K, V>, key: K, getValue: () => V): void {
+        if (!map.has(key)) {
+            map.set(key, getValue());
+        }
+    }
+
+    /** Sets `map.get(key) === getValue()` unless `map.get(key)` was already defined. */
+    export function getOrUpdateMap<K, V>(map: MapCommon<K, V>, key: K, getValue: () => V): V {
+        const value = map.get(key);
+        if (value === undefined) {
+            const value = getValue();
+            map.set(key, value);
+            return value;
+        }
+        else {
+            return value;
+        }
+    }
+
+    /** True iff the predicate is true for some entry in the map. */

```

Comment:
I would still use the whole phrase `if and only if`, if only to avoid unneeded technical jargon. (The terseness afforded by the abbreviation isn't valuable here)